server/src/org/apache/cloudstack/acl/RoleManagerImpl.java

```diff
@@ -0,0 +1,273 @@
+        if (accounts == null || accounts.size() == 0) {
+            return Transaction.execute(new TransactionCallback<Boolean>() {
+                @Override
+                public Boolean doInTransaction(TransactionStatus status) {
+                    List<? extends RolePermission> rolePermissions = rolePermissionsDao.findAllByRoleId(role.getId());
+                    if (rolePermissions != null && !rolePermissions.isEmpty()) {
+                        for (RolePermission rolePermission : rolePermissions) {
+                            rolePermissionsDao.remove(rolePermission.getId());
+                        }
+                    }
+                    return roleDao.remove(role.getId());
+                }
+            });
+        }
+        throw new PermissionDeniedException("Found accounts that have role in use, won't allow to delete role");
+    }
+
+    @Override
+    @ActionEvent(eventType = EventTypes.EVENT_ROLE_PERMISSION_CREATE, eventDescription = "creating Role Permission")
+    public RolePermission createRolePermission(final Role role, final Rule rule, final RolePermission.Permission permission, final String description) {
+        checkCallerAccess();
+        return Transaction.execute(new TransactionCallback<RolePermissionVO>() {
+            @Override
+            public RolePermissionVO doInTransaction(TransactionStatus status) {
+                return rolePermissionsDao.persist(new RolePermissionVO(role.getId(), rule.toString(), permission, description));
+            }
+        });
+    }
+
+    @Override
+    @ActionEvent(eventType = EventTypes.EVENT_ROLE_PERMISSION_UPDATE, eventDescription = "updating Role Permission")
+    public boolean updateRolePermission(final RolePermission rolePermission, final Rule rule, final RolePermission.Permission permission, final String description) {
+        checkCallerAccess();
+        if (rolePermission == null) {
+            return false;
+        }
+        RolePermissionVO rolePermissionVO = (RolePermissionVO) rolePermission;
+        if (rule != null) {
+            rolePermissionVO.setRule(rule.toString());
+        }
+        if (permission != null) {
+            rolePermissionVO.setPermission(permission);
+        }
+        if (!Strings.isNullOrEmpty(description)) {
+            rolePermissionVO.setDescription(description);
+        }
+        return rolePermissionsDao.update(rolePermission.getId(), rolePermissionVO);
+    }
+
+    @Override
+    @ActionEvent(eventType = EventTypes.EVENT_ROLE_PERMISSION_DELETE, eventDescription = "deleting Role Permission")
+    public boolean deleteRolePermission(final RolePermission rolePermission) {
+        checkCallerAccess();
+        return rolePermission != null && rolePermissionsDao.remove(rolePermission.getId());
+    }
+
+    private List<Role> toRoleList(final List<? extends Role> roles) {
+        if (roles != null) {
+            return new ArrayList<>(roles);
+        }
+        return new ArrayList<>();
```

Comment:
Minor nit: ``Collection.emptyList()`` avoids unnecessary object creation in the empty case.